### ISSUE TITLE
Can we please have a version in this tool?

### ISSUE THREAD

**Issue by wolfeidau** [constabulary/gb#212]
Currently there is no version information provided so it is difficult to assert whether the tool needs updating.

As in `gb version` just helps me keep things in sync across machines.

Thanks

**Comment by davecheney** [comment]
Yup, that is a fair request.

At the moment we're moving a orbital speeds, hopefully we'll slow down to a pace where milestones and releases start to make good sense.

What will most likely happen is we'll move from go get to self hosted, and at that point ship actual releases via github.

I can't promise to do this in the short term, but I understand the requirement and think it valid.

**Comment by leifmadsen** [comment]
@davecheney Just cruising by. How much more is required to close this out now that you're tagging releases like 0.1.0 and 0.1.1 on GitHub?

**Comment by davecheney** [comment]
Unfortunately a lot as the go tool does not have the ability to consume a
git tag and annotate a binary. This can be done with extra scripting, but
that won't help people that just use go get.

**Comment by mpasternacki** [comment]
How about scripting the release process? Let's say, a `release.sh` script, which would take version number as argument, and do something along the lines of:

```sh
#!/bin/sh
set -e
VERSION="$1"

cat >version.go <<EOF
package gb

const Version = "$VERSION"
EOF

git commit -m "Release version $VERSION" version.go

# If relevant, sign the tag with gpg
git tag -a -m "Tag version $VERSION" v$VERSION

cat >version.go <<EOF
package gb

const Version = "$VERSION+git"
EOF

git commit -m "Set development version $VERSION+git" version.go

git push --tags
```

This would take care of keeping tags and committed version number in sync. This would mean that building directly from a tagged release (e.g. by a packaging tool) has exact version number defined, and using `go get` directly off master is marked `+git`, so that work-in-progress versions are clearly distinguished. The downside is always having `+git` on `go get`-derived libraries, because master would always be younger than the tag.

A more stable/reliable option could be adopting something like git-flow, where master branch is always latest released version, and a separate develop branch is release-ready development. Then master would always be a tagged revision, develop would add `+git` suffix, and release process would drop it. This is more complex and not everybody likes git-flow, but it has the benefit of master (pulled by default by `go get`) always being the most recent released version.

**Comment by mpasternacki** [comment]
(Hm. Unless `go get` pulls the default branch rather than specifically master. With git-flow-like setup you need to set _develop_ as main branch on github, otherwise all the PRs you'll get will be against master)

**Comment by davecheney** [comment]
Thank you for this feedback. I want to be clear that I am no dismissing your ideas, having a usable version number is very important to me as well.

However, at this early stage in development, I am not planning on addressing this.

Usable internal version numbers is something that I will address as we get closer to a 1.0 release. FOr now, my advise is to always use the latest available.

**Comment by mpasternacki** [comment]
Thank you for your honest reply. Would you be interested in a PR which addresses this, or do you prefer to keep this issue open until later to avoid locking yourself into a particular solution? I'd be willing to prepare a PR if it's just a matter of time needed for implementation.

I'm interested in this as I'm maintaining FreeBSD port for gb (which I use to build my software that I want to package, and packaging the build tool is prerefable to vendoring it). Once a packaged version is available (and it is already – right now the port is at 0.1.0, with update for 0.3.1 waiting for a committer to merge), the "latest available" is not exactly clear, so easy way to check what is the version that actually runs gets more important.

That said, I understand if you prefer to keep it unsolved to avoid maintenance overhead, even if you get a PR. Just let me know if/when you need any help with this feature.

**Comment by davecheney** [comment]
> Thank you for your honest reply. Would you be interested in a PR which addresses this, or do you prefer to keep this issue open until later to avoid locking yourself into a particular solution? I'd be willing to prepare a PR if it's just a matter of time needed for implementation.

I would prefer to not take a PR until gb is closer to a 1.0 release.

> I'm interested in this as I'm maintaining FreeBSD port for gb (which I use to build my software that I want to package, and packaging the build tool is prerefable to vendoring it). Once a packaged version is available (and it is already – right now the port is at 0.1.0, with update for 0.3.1 waiting for a committer to merge), the "latest available" is not exactly clear, so easy way to check what is the version that actually runs gets more important.

Thank you for doing this. Is a git tag sufficient for ports maintainers ?

**Comment by mpasternacki** [comment]
Yes, github tag is sufficient to build a particular version in a repeatable way. Version subcommand would be just extra convenience to confirm which version actually runs. I understand now how it fits into your development priorities and will wait patiently for it :)

**Comment by kalafut** [comment]
Here's one approach the user can add to their .bashrc until the version is baked in:

```
$ alias gbv="(cd \$GOPATH/src/github.com/constabulary/gb && git describe --abbrev=0 HEAD)"
$ gbv
v0.3.1
```

**Comment by davecheney** [comment]
Thanks for the feedback. To give a little more background, with gb
iterating so fast, and considering that the go tool doesn't care about
versions it is expected that a lot of people will not be using a released
version.

Thanks utility of a gb version command is to return the hash their tool was
built from but the go tool provides no mechanism for this.

This could be done with a build script, but then there are three ways of
building gb, with the go tool, with gb, and with a build script, what a
mess.

My intention is to address this problem when gb becomes self hosting, but I
don't have a timeframe for that.

Thanks

Dave

**Comment by freeformz** [comment]
@davecheney FWIW: If you did releases via travis you could extract the git tag and pass that to the go tool to set a string via -ldflags.

I totally get this makes things either harder or easier depending on your pov and totally understand you not wanting to support this until you are closer to 1.0.

**Comment by davecheney** [comment]
Yup, basically I know how to do it, and the costs, and I'm punting on them
til gb gets closer to 1.0
